File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[][]}
def level_order(root)
@result=[]
    def level_traverse level
        return nil if level.empty?
        @result.push(level.map {|node| node.val})
        next_level=[]
        level.each do |node| 
            next_level.push(node.left) if node.left
            next_level.push(node.right) if node.right
        end
        level_traverse next_level
    end
    level_traverse [root] if root
    return @result
end

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    let result=[]
    if(!root)
        return result
    function traverse(nodes){
        if(!nodes.length)
            return
        result.push(nodes.map(node=>node.val))
        let next=[]
        nodes.forEach(node=>{
            if(node.left)
                next.push(node.left)
            if(node.right)
                next.push(node.right)
        })
        traverse(next)
    }
    traverse([root])
    return result 
};